This is the error: 

```
> Task :compileJava NO-SOURCE
> Task :compileGroovy FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileGroovy'.
> io/micronaut/http/client/annotation/Client

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1s
1 actionable task: 1 executed

org.gradle.testkit.runner.UnexpectedBuildFailure: Unexpected build execution failure in /var/folders/65/67w3r16d2cn0jmpygf_mgch80000gn/T/test-createcontroller-createcontrollergroovygradlejunitspec3735964372924414578 with arguments [test, --build-cache]

Output:
> Task :compileJava NO-SOURCE
> Task :compileGroovy FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileGroovy'.
> io/micronaut/http/client/annotation/Client

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1s
1 actionable task: 1 executed

	at app//org.gradle.testkit.runner.internal.DefaultGradleRunner.lambda$build$2(DefaultGradleRunner.java:269)
	at app//org.gradle.testkit.runner.internal.DefaultGradleRunner.run(DefaultGradleRunner.java:362)
	at app//org.gradle.testkit.runner.internal.DefaultGradleRunner.build(DefaultGradleRunner.java:267)
	at app//io.micronaut.starter.test.CommandSpec.executeGradle(CommandSpec.groovy:86)
	at app//io.micronaut.starter.test.CommandSpec.executeBuild(CommandSpec.groovy:77)
	at io.micronaut.starter.cli.test.CreateControllerSpec.test creating a controller and running the test for #language and #testFramework and #buildTool(CreateControllerSpec.groovy:38)
```

--- a/starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryCore.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryCore.java
@@ -26,7 +26,8 @@ import jakarta.inject.Singleton;
 
 @Singleton
 public class DiscoveryCore implements Feature {
-    private static final String ARTIFACT_ID_MICRONAUT_DISCOVERY_CORE = "micronaut-discovery-core";
+
+    public static final String ARTIFACT_ID_MICRONAUT_DISCOVERY_CORE = "micronaut-discovery-core";
     private static final Dependency DEPENDENCY_MICRONAUT_DISCOVERY_CORE = MicronautDependencyUtils.coreDependency()
             .artifactId(ARTIFACT_ID_MICRONAUT_DISCOVERY_CORE)
             .compile()

--- a/starter-core/src/main/java/io/micronaut/starter/feature/k8s/KubernetesClient.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/k8s/KubernetesClient.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017-2021 original authors
+ * Copyright 2017-2023 original authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -19,8 +19,12 @@ import io.micronaut.core.annotation.NonNull;
 import io.micronaut.starter.application.ApplicationType;
 import io.micronaut.starter.application.generator.GeneratorContext;
 import io.micronaut.starter.build.dependencies.Dependency;
+import io.micronaut.starter.build.dependencies.MicronautDependencyUtils;
 import io.micronaut.starter.feature.Category;
 import io.micronaut.starter.feature.Feature;
+import io.micronaut.starter.feature.discovery.DiscoveryCore;
+import io.micronaut.starter.options.BuildTool;
+import io.micronaut.starter.options.Language;
 import jakarta.inject.Singleton;
 
 /**
@@ -76,5 +80,18 @@ public class KubernetesClient implements Feature {
                 .groupId(MICRONAUT_KUBERNETES_GROUP_ID)
                 .artifactId("micronaut-kubernetes-client")
                 .compile());
+        fixupDependencies(generatorContext);
+    }
+
+    static void fixupDependencies(GeneratorContext generatorContext) {
+        if (!generatorContext.hasFeature(DiscoveryCore.class)
+                && generatorContext.getBuildTool() == BuildTool.MAVEN
+                && generatorContext.getLanguage() == Language.GROOVY
+        ) {
+            // Maven requires discovery core provided to work with http-validation under groovy
+            generatorContext.addDependency(MicronautDependencyUtils.coreDependency()
+                    .artifactId(DiscoveryCore.ARTIFACT_ID_MICRONAUT_DISCOVERY_CORE)
+                    .compileOnly());
+        }
     }
 }

--- a/starter-core/src/main/java/io/micronaut/starter/feature/k8s/KubernetesReactorClient.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/k8s/KubernetesReactorClient.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017-2021 original authors
+ * Copyright 2017-2023 original authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -89,5 +89,7 @@ public class KubernetesReactorClient implements Feature {
                 .groupId(KubernetesClient.MICRONAUT_KUBERNETES_GROUP_ID)
                 .artifactId("micronaut-kubernetes-client-reactor")
                 .compile());
+
+        KubernetesClient.fixupDependencies(generatorContext);
     }
 }

--- a/starter-core/src/main/java/io/micronaut/starter/feature/k8s/KubernetesRxJava2Client.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/k8s/KubernetesRxJava2Client.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017-2021 original authors
+ * Copyright 2017-2023 original authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -89,5 +89,7 @@ public class KubernetesRxJava2Client implements Feature {
                 .groupId(KubernetesClient.MICRONAUT_KUBERNETES_GROUP_ID)
                 .artifactId("micronaut-kubernetes-client-rxjava2")
                 .compile());
+
+        KubernetesClient.fixupDependencies(generatorContext);
     }
 }

--- a/starter-core/src/main/java/io/micronaut/starter/feature/other/HttpClientTest.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/other/HttpClientTest.java
@@ -71,7 +71,7 @@ public class HttpClientTest implements DefaultFeature {
             generatorContext.addDependency(HttpClient.DEPENDENCY_MICRONAUT_HTTP_CLIENT);
         } else if (generatorContext.getApplicationType() == ApplicationType.DEFAULT) {
             generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_TEST);
-            if (generatorContext.hasFeature(MicronautHttpValidation.class)) {
+            if (generatorContext.hasFeature(MicronautHttpValidation.class) && generatorContext.getBuildTool().isGradle()) {
                 generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_COMPILE_ONLY);
             }
         }

--- a/starter-core/src/main/java/io/micronaut/starter/feature/other/HttpClientTest.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/other/HttpClientTest.java
@@ -25,6 +25,7 @@ import io.micronaut.starter.feature.awslambdacustomruntime.AwsLambdaCustomRuntim
 import io.micronaut.starter.feature.function.awslambda.AwsLambda;
 import io.micronaut.starter.feature.graalvm.GraalVM;
 import io.micronaut.starter.feature.httpclient.HttpClientFeature;
+import io.micronaut.starter.feature.validator.MicronautHttpValidation;
 import io.micronaut.starter.options.Options;
 import jakarta.inject.Singleton;
 import java.util.Set;
@@ -33,9 +34,15 @@ import static io.micronaut.starter.feature.other.HttpClient.ARTIFACT_ID_MICRONAU
 
 @Singleton
 public class HttpClientTest implements DefaultFeature {
+
     private static final Dependency DEPENDENCY_MICRONAUT_HTTP_CLIENT_TEST = MicronautDependencyUtils.coreDependency()
             .artifactId(ARTIFACT_ID_MICRONAUT_HTTP_CLIENT)
             .test()
+            .build();
+
+    private static final Dependency DEPENDENCY_MICRONAUT_HTTP_CLIENT_COMPILE_ONLY = MicronautDependencyUtils.coreDependency()
+            .artifactId(ARTIFACT_ID_MICRONAUT_HTTP_CLIENT)
+            .compileOnly()
             .build();
 
     @Override
@@ -64,6 +71,9 @@ public class HttpClientTest implements DefaultFeature {
             generatorContext.addDependency(HttpClient.DEPENDENCY_MICRONAUT_HTTP_CLIENT);
         } else if (generatorContext.getApplicationType() == ApplicationType.DEFAULT) {
             generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_TEST);
+            if (generatorContext.hasFeature(MicronautHttpValidation.class)) {
+                generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_CLIENT_COMPILE_ONLY);
+            }
         }
     }
 }

--- a/starter-core/src/main/java/io/micronaut/starter/feature/validator/MicronautHttpValidation.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/validator/MicronautHttpValidation.java
@@ -71,6 +71,7 @@ public class MicronautHttpValidation implements DefaultFeature {
     protected void addDependencies(GeneratorContext generatorContext) {
         generatorContext.addDependency(DEPENDENCY_MICRONAUT_HTTP_VALIDATION);
     }
+
     @Override
     public boolean shouldApply(ApplicationType applicationType, Options options, Set<Feature> selectedFeatures) {
         return applicationType == ApplicationType.DEFAULT;

--- a/test-cli/src/test/groovy/io/micronaut/starter/cli/test/CreateControllerSpec.groovy
+++ b/test-cli/src/test/groovy/io/micronaut/starter/cli/test/CreateControllerSpec.groovy
@@ -9,11 +9,9 @@ import io.micronaut.starter.options.Language
 import io.micronaut.starter.options.TestFramework
 import io.micronaut.starter.test.CommandSpec
 import io.micronaut.starter.test.LanguageBuildTestFrameworkCombinations
-import spock.lang.Unroll
 
 class CreateControllerSpec extends CommandSpec {
 
-    @Unroll
     void "test creating a controller and running the test for #language and #testFramework and #buildTool"(Language language,
                                                                                                            BuildTool buildTool,
                                                                                                            TestFramework testFramework) {

--- a/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/k8s/KubernetesClientSpec.groovy
+++ b/test-features/src/test/groovy/io/micronaut/starter/core/test/feature/k8s/KubernetesClientSpec.groovy
@@ -6,7 +6,6 @@ import io.micronaut.starter.test.BuildToolTest
 import io.micronaut.starter.test.CommandSpec
 import org.gradle.testkit.runner.BuildResult
 import spock.lang.IgnoreIf
-import spock.lang.Unroll
 
 class KubernetesClientSpec extends CommandSpec {
 
@@ -16,7 +15,6 @@ class KubernetesClientSpec extends CommandSpec {
     }
 
     @IgnoreIf({ BuildToolTest.IGNORE_MAVEN })
-    @Unroll
     void "test maven #feature with #language"(String feature, Language language) {
         when:
         generateProject(language, BuildTool.MAVEN, [feature])
@@ -31,7 +29,6 @@ class KubernetesClientSpec extends CommandSpec {
                 Language.values()].combinations()
     }
 
-    @Unroll
     void "test gradle #feature with #language"(String feature, Language language) {
         when:
         generateProject(language, BuildTool.GRADLE, [feature])

